File: typescript/src/utils/structured-logger.ts
export type LogLevel = "DEBUG" | "INFO" | "WARN" | "ERROR";

const LOG_LEVELS: Record<LogLevel, number> = {
  DEBUG: 0,
  INFO: 1,
  WARN: 2,
  ERROR: 3,
};

export interface StructuredLoggerConfig {
  minLevel?: LogLevel;
  service?: string;
  prettyPrint?: boolean;
}

export class StructuredLogger {
  private minLevel: number;
  private service: string;
  private pretty: boolean;

  constructor(config?: StructuredLoggerConfig) {
    this.minLevel = config?.minLevel ? LOG_LEVELS[config.minLevel] : LOG_LEVELS.INFO;
    this.service = config?.service ?? "mpesa-sdk";
    this.pretty = config?.prettyPrint ?? false;
  }

  private log(level: LogLevel, msg: string, meta?: Record<string, unknown>): void {
    if (LOG_LEVELS[level] < this.minLevel) return;

    const entry: Record<string, unknown> = {
      timestamp: new Date().toISOString(),
      level,
      message: msg,
      service: this.service,
    };

    if (meta && Object.keys(meta).length > 0) {
      entry.metadata = meta;
    }

    const output = this.pretty ? JSON.stringify(entry, null, 2) : JSON.stringify(entry);

    switch (level) {
      case "ERROR":
        console.error(output);
        break;
      case "WARN":
        console.warn(output);
        break;
      case "DEBUG":
        console.debug(output);
        break;
      default:
        console.log(output);
    }
  }

  debug(msg: string, meta?: Record<string, unknown>): void {
    this.log("DEBUG", msg, meta);
  }

  info(msg: string, meta?: Record<string, unknown>): void {
    this.log("INFO", msg, meta);
  }

  warn(msg: string, meta?: Record<string, unknown>): void {
    this.log("WARN", msg, meta);
  }

  error(msg: string, meta?: Record<string, unknown>): void {
    this.log("ERROR", msg, meta);
  }

  child(service: string): StructuredLogger {
    return new StructuredLogger({
      minLevel: this.minLevel as unknown as LogLevel,
      service: `${this.service}.${service}`,
      prettyPrint: this.pretty,
    });
  }
}

File: packages/python/src/daraja/utils/structured_logger.py
import json
import logging
import sys
from datetime import UTC, datetime
from typing import Any


class StructuredLogHandler(logging.Handler):
    def __init__(self, stream=None, pretty: bool = False):
        super().__init__()
        self.stream = stream or sys.stderr
        self.pretty = pretty

    def emit(self, record: logging.LogRecord):
        entry = {
            "timestamp": datetime.fromtimestamp(record.created, tz=UTC).isoformat(),
            "level": record.levelname,
            "message": record.getMessage(),
            "service": "mpesa-sdk",
            "logger": record.name,
        }

        if hasattr(record, "extra_fields") and record.extra_fields:
            entry["metadata"] = record.extra_fields

        if self.pretty:
            line = json.dumps(entry, indent=2, default=str)
        else:
            line = json.dumps(entry, default=str)

        self.stream.write(line + "\n")
        self.flush()


class StructuredLogger:
    def __init__(self, min_level: str = "INFO", service: str = "mpesa-sdk", pretty: bool = False):
        self.service = service
        self.pretty = pretty

        level_map = {
            "DEBUG": logging.DEBUG,
            "INFO": logging.INFO,
            "WARN": logging.WARNING,
            "WARNING": logging.WARNING,
            "ERROR": logging.ERROR,
        }
        self._logger = logging.getLogger(service)
        self._logger.setLevel(level_map.get(min_level.upper(), logging.INFO))
        self._logger.handlers.clear()
        self._logger.addHandler(StructuredLogHandler(pretty=pretty))
        self._logger.propagate = False

    def _log(self, level: int, msg: str, **kwargs: Any):
        extra = kwargs.pop("extra", None) or {}
        metadata = kwargs.pop("metadata", None) or extra
        if "request_id" in metadata:
            metadata["requestId"] = metadata.pop("request_id")
        record = self._logger.makeRecord(
            self._logger.name, level, "", 0, msg, (), None,
        )
        record.extra_fields = metadata if metadata else {}
        self._logger.handle(record)

    def debug(self, msg: str, *args: Any, **kwargs: Any) -> None:
        if args:
            msg = msg % args
        self._log(logging.DEBUG, msg, **kwargs)

    def info(self, msg: str, *args: Any, **kwargs: Any) -> None:
        if args:
            msg = msg % args
        self._log(logging.INFO, msg, **kwargs)

    def warning(self, msg: str, *args: Any, **kwargs: Any) -> None:
        if args:
            msg = msg % args
        self._log(logging.WARNING, msg, **kwargs)

    def error(self, msg: str, *args: Any, **kwargs: Any) -> None:
        if args:
            msg = msg % args
        self._log(logging.ERROR, msg, **kwargs)
